File: PythonArchive/Shortest-Peak-Reversal-5m-3x-Margin--main/Unity App Project/short_trader_multi_filter/Assets/Scripts/Core/BacktestEngine.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace UnityApp.ShortTraderMultiFilter
{
    public class BacktestEngine
    {
        private readonly TraderConfig _config;

        public BacktestEngine(TraderConfig config)
        {
            _config = config;
        }

        public BacktestMetrics RunBacktest(IReadOnlyList<Candle> candles, StrategyParams parameters, bool captureTrades = false)
        {
            var ordered = candles.OrderBy(c => c.Timestamp).ToList();
            var closes = ordered.Select(c => c.Close).ToList();
            var highs = ordered.Select(c => c.High).ToList();
            var lows = ordered.Select(c => c.Low).ToList();

            var sma = RollingMean(closes, parameters.SmaPeriod);
            var lowestLow = RollingMin(lows, parameters.StochPeriod);
            var highestHigh = RollingMax(highs, parameters.StochPeriod);

            var rawStoch = new List<double?>();
            for (var i = 0; i < ordered.Count; i++)
            {
                if (!lowestLow[i].HasValue || !highestHigh[i].HasValue || Math.Abs(highestHigh[i]!.Value - lowestLow[i]!.Value) < 1e-12)
                {
                    rawStoch.Add(0);
                }
                else
                {
                    var value = 100 * (closes[i] - lowestLow[i]!.Value) / (highestHigh[i]!.Value - lowestLow[i]!.Value);
                    rawStoch.Add(value);
                }
            }

            var smoothedK = RollingMean(rawStoch, _config.SmoothK).Select(x => x.HasValue ? x - 50 : (double?)null).ToList();
            var emaFast = ExponentialMovingAverage(closes, parameters.MacdFast);
            var emaSlow = ExponentialMovingAverage(closes, parameters.MacdSlow);
            var macd = emaFast.Zip(emaSlow, (fast, slow) => fast - slow).ToList();
            var signal = ExponentialMovingAverage(macd, parameters.MacdSignal);

            var balance = _config.StartingBalance;
            var equityCurve = new List<double>();
            var trades = new List<TradeRecord>();
            PositionState? position = null;
            var wins = 0;
            var losses = 0;
            var winSizes = new List<double>();
            var lossSizes = new List<double>();

            var warmup = Math.Max(parameters.SmaPeriod, Math.Max(parameters.StochPeriod, Math.Max(parameters.MacdSlow, parameters.MacdSignal))) + 2;
            for (var i = warmup; i < ordered.Count; i++)
            {
                var candle = ordered[i];
                var close = candle.Close;

                var inDate = candle.Timestamp.Year > _config.StartYear ||
                             (candle.Timestamp.Year == _config.StartYear && candle.Timestamp.Month >= _config.StartMonth);

                if (position == null)
                {
                    if (!inDate)
                    {
                        equityCurve.Add(balance);
                        continue;
                    }

                    var lowsOk = ordered[i - 2].Low <= ordered[i - 1].Low && candle.Low < ordered[i - 1].Low;
                    var smaOk = sma[i].HasValue && sma[i - 1].HasValue && sma[i] < sma[i - 1];
                    var macdOk = !parameters.UseMacd || (macd[i] < macd[i - 1]);
                    var signalOk = !parameters.UseSignal || (signal[i].HasValue && signal[i - 1].HasValue && signal[i] < signal[i - 1]);

                    if (lowsOk && smaOk && macdOk && signalOk)
                    {
                        var riskFraction = _config.RiskFraction;
                        var marginRate = _config.MarginRate;
                        var positionValue = (balance * riskFraction) / marginRate;
                        var qty = positionValue / close;
                        var marginUsed = balance * riskFraction;
                        if (marginUsed <= 0 || qty <= 0)
                        {
                            equityCurve.Add(balance);
                            continue;
                        }

                        balance -= marginUsed;
                        var tp = close * (1 - 0.004);
                        position = new PositionState
                        {
                            Side = "short",
                            EntryPrice = close,
                            TakeProfit = tp,
                            Quantity = qty,
                            EntryTime = candle.Timestamp,
                            MarginUsed = marginUsed,
                        };

                        equityCurve.Add(balance + marginUsed);
                        continue;
                    }
                }
                else
                {
                    var tpHit = position.TakeProfit.HasValue && candle.Low <= position.TakeProfit.Value;
                    var momentumExit = parameters.UseMomentumExit &&
                                       smoothedK[i].HasValue && smoothedK[i - 1].HasValue &&
                                       smoothedK[i] > smoothedK[i - 1];

                    double? exitPrice = null;
                    var exitType = "momentum";
                    if (tpHit)
                    {
                        exitPrice = position.TakeProfit;
                        exitType = "tp";
                    }
                    else if (momentumExit)
                    {
                        exitPrice = close;
                    }

                    if (exitPrice.HasValue)
                    {
                        var gross = (position.EntryPrice - exitPrice.Value) * position.Quantity;
                        balance += position.MarginUsed + gross;
                        var pnlPct = (gross / _config.StartingBalance) * 100;
                        if (gross > 0)
                        {
                            wins++;
                            winSizes.Add(pnlPct);
                        }
                        else
                        {
                            losses++;
                            lossSizes.Add(pnlPct);
                        }

                        if (captureTrades)
                        {
                            trades.Add(new TradeRecord(
                                position.EntryTime,
                                candle.Timestamp,
                                "SHORT",
                                position.EntryPrice,
                                exitPrice.Value,
                                gross,
                                pnlPct,
                                position.Quantity,
                                exitType));
                        }

                        position = null;
                    }
                }

                equityCurve.Add(balance + (position?.MarginUsed ?? 0));
            }

            if (position != null)
            {
                var finalClose = ordered[^1].Close;
                var gross = (position.EntryPrice - finalClose) * position.Quantity;
                balance += position.MarginUsed + gross;
                var pnlPct = (gross / _config.StartingBalance) * 100;
                if (gross > 0)
                {
                    wins++;
                    winSizes.Add(pnlPct);
                }
                else
                {
                    losses++;
                    lossSizes.Add(pnlPct);
                }
                if (captureTrades)
                {
                    trades.Add(new TradeRecord(
                        position.EntryTime,
                        ordered[^1].Timestamp,
                        "SHORT",
                        position.EntryPrice,
                        finalClose,
                        gross,
                        pnlPct,
                        position.Quantity,
                        "final_close"));
                }
                equityCurve.Add(balance);
            }

            if (equityCurve.Count == 0)
            {
                return new BacktestMetrics(0, 0, _config.StartingBalance, 0, 0, 0, null, 0, 0, 0, trades);
            }

            var finalBalance = equityCurve[^1];
            var pnlValue = finalBalance - _config.StartingBalance;
            var pnlPctTotal = (pnlValue / _config.StartingBalance) * 100;
            var avgWin = winSizes.Count > 0 ? winSizes.Average() : 0;
            var avgLoss = lossSizes.Count > 0 ? lossSizes.Average() : 0;
            var winRate = (wins + losses) > 0 ? (double)wins / (wins + losses) * 100 : 0;
            double? rr = avgLoss != 0 ? avgWin / Math.Abs(avgLoss) : null;

            var returns = new List<double>();
            for (var i = 1; i < equityCurve.Count; i++)
            {
                var prev = equityCurve[i - 1];
                var curr = equityCurve[i];
                if (prev != 0)
                {
                    returns.Add((curr - prev) / prev);
                }
            }

            var meanReturn = returns.Count > 0 ? returns.Average() : 0;
            var stdReturn = returns.Count > 1 ? StdDev(returns) : 0;
            var sharpe = stdReturn != 0 ? meanReturn / stdReturn * Math.Sqrt(365 * 24 * 60 / _config.AggregationMinutes) : 0;

            return new BacktestMetrics(pnlPctTotal, pnlValue, finalBalance, avgWin, avgLoss, winRate, rr, sharpe, wins, losses, trades);
        }

        public IReadOnlyList<Dictionary<string, object>> GridSearch(IReadOnlyList<Candle> candles)
        {
            var results = new List<Dictionary<string, object>>();

            foreach (var sma in _config.SmaPeriodRange)
            {
                foreach (var stoch in _config.StochPeriodRange)
                {
                    foreach (var useMacd in _config.UseMacdOptions)
                    {
                        foreach (var useSignal in _config.UseSignalOptions)
                        {
                            foreach (var useMom in _config.UseMomentumExitOptions)
                            {
                                var parameters = new StrategyParams(
                                    sma,
                                    stoch,
                                    _config.MacdFast,
                                    _config.MacdSlow,
                                    _config.MacdSignal,
                                    useMacd,
                                    useSignal,
                                    useMom);

                                var metrics = RunBacktest(candles, parameters, false);
                                var row = new Dictionary<string, object>
                                {
                                    ["sma_period"] = sma,
                                    ["stoch_period"] = stoch,
                                    ["macd_fast"] = _config.MacdFast,
                                    ["macd_slow"] = _config.MacdSlow,
                                    ["macd_signal"] = _config.MacdSignal,
                                    ["use_macd"] = useMacd,
                                    ["use_signal"] = useSignal,
                                    ["use_momentum_exit"] = useMom,
                                    ["pnl_pct"] = metrics.PnlPct,
                                    ["pnl_value"] = metrics.PnlValue,
                                    ["final_balance"] = metrics.FinalBalance,
                                    ["wins"] = metrics.Wins,
                                    ["losses"] = metrics.Losses,
                                    ["avg_win"] = metrics.AverageWinPct,
                                    ["avg_loss"] = metrics.AverageLossPct,
                                    ["win_rate"] = metrics.WinRatePct,
                                    ["rr_ratio"] = metrics.RiskReward,
                                    ["sharpe"] = metrics.Sharpe
                                };
                                results.Add(row);
                            }
                        }
                    }
                }
            }

            return results;
        }

        public static Dictionary<string, double> SummarizeResults(Dictionary<string, object> bestRow, double startingBalance)
        {
            var wins = Convert.ToInt32(bestRow["wins"]);
            var losses = Convert.ToInt32(bestRow["losses"]);
            var totalTrades = wins + losses;
            var finalBalance = Convert.ToDouble(bestRow["final_balance"]);
            var totalPnl = Convert.ToDouble(bestRow["pnl_value"]);
            var avgWin = Convert.ToDouble(bestRow["avg_win"]);
            var avgLoss = Convert.ToDouble(bestRow["avg_loss"]);

            var winRate = totalTrades > 0 ? (double)wins / totalTrades * 100 : 0;

            return new Dictionary<string, double>
            {
                ["Total Trades"] = totalTrades,
                ["Wins"] = wins,
                ["Losses"] = losses,
                ["Win Rate %"] = Math.Round(winRate, 2),
                ["Total PnL"] = Math.Round(totalPnl, 2),
                ["Final Balance"] = Math.Round(finalBalance, 2),
                ["Average Win"] = Math.Round(avgWin, 2),
                ["Average Loss"] = Math.Round(avgLoss, 2)
            };
        }

        private static List<double?> RollingMean(IReadOnlyList<double> values, int period)
        {
            var result = new List<double?>(new double?[values.Count]);
            if (period <= 0)
            {
                return result;
            }
            double sum = 0;
            for (var i = 0; i < values.Count; i++)
            {
                sum += values[i];
                if (i >= period)
                {
                    sum -= values[i - period];
                }
                if (i >= period - 1)
                {
                    result[i] = sum / period;
                }
            }
            return result;
        }

        private static List<double?> RollingMean(IReadOnlyList<double?> values, int period)
        {
            var result = new List<double?>(new double?[values.Count]);
            if (period <= 0)
            {
                return result;
            }

            var window = new Queue<double>();
            for (var i = 0; i < values.Count; i++)
            {
                var val = values[i];
                if (val.HasValue)
                {
                    window.Enqueue(val.Value);
                }

                if (window.Count > period)
                {
                    window.Dequeue();
                }

                if (window.Count == period)
                {
                    result[i] = window.Average();
                }
            }

            return result;
        }

        private static List<double?> RollingMin(IReadOnlyList<double> values, int period)
        {
            var result = new List<double?>(new double?[values.Count]);
            if (period <= 0)
            {
                return result;
            }
            for (var i = 0; i < values.Count; i++)
            {
                if (i >= period - 1)
                {
                    var window = values.Skip(i - period + 1).Take(period);
                    result[i] = window.Min();
                }
            }
            return result;
        }

        private static List<double?> RollingMax(IReadOnlyList<double> values, int period)
        {
            var result = new List<double?>(new double?[values.Count]);
            if (period <= 0)
            {
                return result;
            }
            for (var i = 0; i < values.Count; i++)
            {
                if (i >= period - 1)
                {
                    var window = values.Skip(i - period + 1).Take(period);
                    result[i] = window.Max();
                }
            }
            return result;
        }

        private static List<double> ExponentialMovingAverage(IReadOnlyList<double> values, int period)
        {
            var result = new List<double>(new double[values.Count]);
            if (period <= 0 || values.Count == 0)
            {
                return result;
            }

            var multiplier = 2.0 / (period + 1);
            result[0] = values[0];
            for (var i = 1; i < values.Count; i++)
            {
                result[i] = ((values[i] - result[i - 1]) * multiplier) + result[i - 1];
            }
            return result;
        }

        private static double StdDev(IReadOnlyList<double> values)
        {
            var mean = values.Average();
            var variance = values.Select(v => Math.Pow(v - mean, 2)).Average();
            return Math.Sqrt(variance);
        }
    }
}

File: PythonArchive/Shortest-Peak-Reversal-5m-3x-Margin--main/src/short_trader_multi_filter/backtest_engine.py
from __future__ import annotations

from dataclasses import dataclass
from itertools import product
from typing import Dict, List

import numpy as np
import pandas as pd
from tqdm import tqdm

from .config import TraderConfig
from .order_utils import PositionState


@dataclass
class StrategyParams:
    sma_period: int
    stoch_period: int
    macd_fast: int
    macd_slow: int
    macd_signal: int
    use_macd: bool
    use_signal: bool
    use_momentum_exit: bool


@dataclass
class BacktestMetrics:
    pnl_pct: float
    pnl_value: float
    final_balance: float
    avg_win: float
    avg_loss: float
    win_rate: float
    rr_ratio: float | None
    sharpe: float
    drawdown: float
    wins: int
    losses: int


def summarize_results(best_row: pd.DataFrame, starting_balance: float) -> Dict[str, float]:
    l = best_row.iloc[0]
    total_trades = int(l["wins"] + l["losses"])
    total_wins = int(l["wins"])
    total_losses = int(l["losses"])
    total_pnl = float(l["pnl_value"])
    final_balance = float(l["final_balance"])
    win_rate = (total_wins / total_trades) * 100 if total_trades > 0 else 0
    avg_win = float(l["avg_win"])
    avg_loss = float(l["avg_loss"])
    return {
        "Total Trades": total_trades,
        "Wins": total_wins,
        "Losses": total_losses,
        "Win Rate %": round(win_rate, 2),
        "Total PnL": round(total_pnl, 2),
        "Final Balance": round(final_balance, 2),
        "Average Win": round(avg_win, 2),
        "Average Loss": round(avg_loss, 2),
    }


class BacktestEngine:
    def __init__(self, config: TraderConfig):
        self.config = config
        self._last_trades: List[Dict] = []

    def _run_backtest(self, df: pd.DataFrame, params: StrategyParams, capture_trades: bool = False) -> BacktestMetrics:
        data = df.copy().sort_index()
        data["sma"] = data["Close"].rolling(params.sma_period).mean()

        lowest_low = data["Low"].rolling(params.stoch_period).min()
        highest_high = data["High"].rolling(params.stoch_period).max()
        raw_stoch = 100 * (data["Close"] - lowest_low) / (highest_high - lowest_low)
        raw_stoch = raw_stoch.replace([np.inf, -np.inf], np.nan).fillna(0)
        data["k"] = raw_stoch.rolling(self.config.smooth_k).mean() - 50

        data["ema_fast"] = data["Close"].ewm(span=params.macd_fast, adjust=False).mean()
        data["ema_slow"] = data["Close"].ewm(span=params.macd_slow, adjust=False).mean()
        data["macd"] = data["ema_fast"] - data["ema_slow"]
        data["signal"] = data["macd"].ewm(span=params.macd_signal, adjust=False).mean()

        balance = self.config.starting_balance
        equity_curve: List[float] = []
        position: PositionState | None = None
        wins = 0
        losses = 0
        win_sizes: List[float] = []
        loss_sizes: List[float] = []
        trades: List[Dict] = [] if capture_trades else []

        warmup = max(params.sma_period, params.stoch_period, params.macd_slow, params.macd_signal) + 2
        for i in range(warmup, len(data)):
            row = data.iloc[i]
            close = float(row["Close"])

            year = row.name.year
            month = row.name.month
            in_date = (year > self.config.start_year) or (year == self.config.start_year and month >= self.config.start_month)

            if position is None:
                if not in_date:
                    equity_curve.append(balance)
                    continue

                lows_ok = data["Low"].iloc[i - 2] <= data["Low"].iloc[i - 1] and data["Low"].iloc[i] < data["Low"].iloc[i - 1]
                sma_ok = row["sma"] < data["sma"].iloc[i - 1]
                macd_ok = (not params.use_macd) or (row["macd"] < data["macd"].iloc[i - 1])
                signal_ok = (not params.use_signal) or (row["signal"] < data["signal"].iloc[i - 1])

                if lows_ok and sma_ok and macd_ok and signal_ok and not np.isnan(row["sma"]):
                    risk_fraction = self.config.risk_fraction
                    margin_rate = self.config.margin_rate
                    position_value = (balance * risk_fraction) / margin_rate
                    qty = position_value / close
                    margin_used = balance * risk_fraction
                    if margin_used <= 0 or qty <= 0:
                        equity_curve.append(balance)
                        continue

                    balance -= margin_used
                    tp_price = close * (1 - 0.004)
                    position = PositionState(side="short", entry_price=close, tp_price=tp_price, qty=qty, entry_bar_time=row.name, margin_used=margin_used)
                    equity_curve.append(balance + margin_used)
                    continue

            else:
                exit_price: float | None = None
                tp_hit = row["Low"] <= (position.tp_price or 0)
                mom_exit = params.use_momentum_exit and (row["k"] > data["k"].iloc[i - 1])

                if tp_hit:
                    exit_price = float(position.tp_price)
                elif mom_exit:
                    exit_price = close

                if exit_price is not None:
                    gross = (position.entry_price - exit_price) * position.qty  # short PnL
                    balance += position.margin_used + gross
                    pnl_pct = (gross / self.config.starting_balance) * 100
                    if gross > 0:
                        wins += 1
                        win_sizes.append(pnl_pct)
                    else:
                        losses += 1
                        loss_sizes.append(pnl_pct)
                    if capture_trades:
                        trades.append(
                            {
                                "entry_time": position.entry_bar_time,
                                "exit_time": row.name,
                                "side": "SHORT",
                                "entry_price": position.entry_price,
                                "exit_price": exit_price,
                                "pnl_value": gross,
                                "pnl_pct": pnl_pct,
                                "qty": position.qty,
                                "exit_type": "tp" if tp_hit else "momentum",
                            }
                        )
                    position = None

            equity_curve.append(balance + (position.margin_used if position else 0))

        if position is not None:
            # Close any open trade at the last available price rather than force-marking it as a loss.
            final_close = float(data.iloc[-1]["Close"])
            gross = (position.entry_price - final_close) * position.qty
            balance += position.margin_used + gross
            pnl_pct = (gross / self.config.starting_balance) * 100
            if gross > 0:
                wins += 1
                win_sizes.append(pnl_pct)
            else:
                losses += 1
                loss_sizes.append(pnl_pct)
            if capture_trades:
                trades.append(
                    {
                        "entry_time": position.entry_bar_time,
                        "exit_time": data.index[-1],
                        "side": "SHORT",
                        "entry_price": position.entry_price,
                        "exit_price": final_close,
                        "pnl_value": gross,
                        "pnl_pct": pnl_pct,
                        "qty": position.qty,
                        "exit_type": "final_close",
                    }
                )
            equity_curve.append(balance)

        if not equity_curve:
            return BacktestMetrics(0, 0, self.config.starting_balance, 0, 0, 0, None, 0, 0, 0, 0)

        final_balance = equity_curve[-1]
        pnl_value = final_balance - self.config.starting_balance
        pnl_pct = (pnl_value / self.config.starting_balance) * 100
        avg_win = float(np.mean(win_sizes)) if win_sizes else 0
        avg_loss = float(np.mean(loss_sizes)) if loss_sizes else 0
        win_rate = wins / (wins + losses) * 100 if (wins + losses) > 0 else 0
        rr_ratio = (avg_win / abs(avg_loss)) if avg_loss != 0 else None
        returns = pd.Series(equity_curve).pct_change().dropna()
        sharpe = (returns.mean() / returns.std()) * np.sqrt(365 * 24 * 60 / self.config.agg_minutes) if returns.std() != 0 else 0

        self._last_trades = trades if capture_trades else []
        return BacktestMetrics(pnl_pct, pnl_value, final_balance, avg_win, avg_loss, win_rate, rr_ratio, sharpe, 0, wins, losses)

    def run_backtest_with_trades(self, df_1m: pd.DataFrame, params: StrategyParams) -> tuple[pd.DataFrame, pd.DataFrame]:
        metrics = self._run_backtest(df_1m, params, capture_trades=True)
        trades_df = pd.DataFrame(self._last_trades) if hasattr(self, "_last_trades") else pd.DataFrame()
        metrics_df = pd.DataFrame([{**params.__dict__, **metrics.__dict__}])
        return metrics_df, trades_df

    def grid_search_with_progress(self, df_1m: pd.DataFrame) -> pd.DataFrame:
        results: List[Dict] = []
        total = (
            len(self.config.sma_period_range)
            * len(self.config.stoch_period_range)
            * len(self.config.use_macd_options)
            * len(self.config.use_signal_options)
            * len(self.config.use_momentum_exit_options)
        )

        for sma_p, stoch_p, use_macd, use_signal, use_mom in tqdm(
            product(
                self.config.sma_period_range,
                self.config.stoch_period_range,
                self.config.use_macd_options,
                self.config.use_signal_options,
                self.config.use_momentum_exit_options,
            ),
            total=total,
            desc="Param search",
            ncols=80,
        ):
            params = StrategyParams(
                sma_period=int(sma_p),
                stoch_period=int(stoch_p),
                macd_fast=self.config.macd_fast,
                macd_slow=self.config.macd_slow,
                macd_signal=self.config.macd_signal,
                use_macd=bool(use_macd),
                use_signal=bool(use_signal),
                use_momentum_exit=bool(use_mom),
            )
            metrics = self._run_backtest(df_1m, params, capture_trades=False)
            results.append({**params.__dict__, **metrics.__dict__})

        return pd.DataFrame(results)
